### PR TITLE
feature/map

### PR DESCRIPTION
**Pull request by Joost-Robat**

### CODE REVIEW

File: Assets/Prefabs/Popup/Map/TestMapPopup.prefab
Comment (PiterGroot) [inline Assets/Prefabs/Popup/Map/TestMapPopup.prefab:1]: Please apply `_trackingTransform` to the prefab, it's missing.

**Review by PiterGroot — CHANGES_REQUESTED** [review]
Only need to check the map popup prefab, good stuff!

File: Assets/Prefabs/Popup/Map/TestMapPopup.prefab
Comment (Joost-Robat) [reply Assets/Prefabs/Popup/Map/TestMapPopup.prefab:1]: Applied!

File: Assets/Art/2D/Sprites/Temp/S_MapPlaceHolder.JPG
Comment (lucasvdiepen) [inline Assets/Art/2D/Sprites/Temp/S_MapPlaceHolder.JPG:1]: Placeholder is one word. Also change the `.JPG` to lowercase.

File: Assets/Scripts/Framework/Gameplay/Map/MapUpdater.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+namespace DerailedDeliveries.Framework.Gameplay.Map
+{
+    /// <summary>
+    /// A class that is responsible for updating the visuals of the map.
+    /// </summary>
+    public class MapUpdater : MonoBehaviour
+    {
+        [Header("World Transforms")]
+        [SerializeField]
+        private Transform _bottomRightWorld;
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Map/MapUpdater.cs]:
```suggestion
        private Transform _bottomRightWorldTransform;
```

File: Assets/Scripts/Framework/Gameplay/Map/MapUpdater.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
+using UnityEngine;
+
+namespace DerailedDeliveries.Framework.Gameplay.Map
+{
+    /// <summary>
+    /// A class that is responsible for updating the visuals of the map.
+    /// </summary>
+    public class MapUpdater : MonoBehaviour
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Map/MapUpdater.cs]: Add `[RequireComponent]` for `RectTransform`.

**Review by lucasvdiepen — CHANGES_REQUESTED** [review]
The bottom right transform is not applied to the environment prefab.

File: Assets/Scripts/Framework/Gameplay/Map/MapManager.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+            _mapIndicator.position = ReturnLerpPosition(index, lerpAlpha);
+        }
+
+        private Vector3 ReturnLerpPosition(int index, float lerpAlpha)
+        {
+            Vector3 currentPos = _currentTrack.MapPath[index].position;
+            Vector3 endPos = _currentTrack.MapPath[index + 1].position;
+
+            return Vector3.Lerp
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/Gameplay/Map/MapManager.cs]:
```suggestion
            return Vector3.Lerp (currentPos, endPos, lerpAlpha);
```

File: Assets/Scripts/Framework/Gameplay/Level/TrainStation.cs [hunk 4]
```diff
@@ -2,6 +2,7 @@
 using UnityEngine;
 
 using DerailedDeliveries.Framework.UI.TextUpdaters;
+using System;
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/Gameplay/Level/TrainStation.cs]: Move up a bit

File: Assets/Art/2D/Sprites/Temp/S_MapPlaceHolder.JPG
Comment (PiterGroot) [inline Assets/Art/2D/Sprites/Temp/S_MapPlaceHolder.JPG:1]: This can be deleted

**Review by PiterGroot — CHANGES_REQUESTED** [review]
Good stuff.

File: Assets/Art/2D/Sprites/Temp.meta
Comment (lucasvdiepen) [inline Assets/Art/2D/Sprites/Temp.meta:1]: Folder can be removed.

File: Assets/Art/2D/Textures/Train/Interior/T_Table.png
Comment (lucasvdiepen) [inline Assets/Art/2D/Textures/Train/Interior/T_Table.png:1]: Missing suffix.

File: Assets/Scripts/Framework/Gameplay/Map/MapManager.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                trackIndex++;
+
+                _tracks[trackIndex]._badSplitWarning.enabled = !_trainController.BadRailSplitOrder[i];
+                trackIndex++;
+            }
+        }
+
+        private void OnDisable() => _trainController.OnDistanceAlongSplineChanged -= UpdateDistanceAlongSpline;
+
+        private void UpdateTrackID(int newTrackID, bool isBadRailsplit = false)
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Map/MapManager.cs]:
```suggestion
        private void UpdateTrackID(int newTrackID, bool isBadRailSplit = false)
```

File: Assets/Scripts/Framework/Gameplay/Map/MapTrack.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /// <summary>
+    /// A class that holds data and functionality for updating the current position on the map.
+    /// </summary>
+    public class MapTrack : MonoBehaviour
+    {
+        /// <summary>
+        /// The TrackID of this <see cref="MapTrack"/>.
+        /// </summary>
+        public int TrackID = -1;
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Map/MapTrack.cs]:
```suggestion
        public int trackID = -1;
```

File: Assets/Scripts/Framework/Gameplay/Map/MapTrack.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+        /// <summary>
+        /// The <see cref="SpriteRenderer"/> of the bad split warning sprite.
+        /// </summary>
+        public SpriteRenderer _badSplitWarning;
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Map/MapTrack.cs]:
```suggestion
        public SpriteRenderer badSplitWarning;
```

File: Assets/Scripts/Framework/Gameplay/Map/MapTrack.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+        /// <summary>
+        /// The array of <see cref="RectTransform"/>s that is lerped from <see cref="RectTransform"/> to 
+        /// <see cref="RectTransform"/> to represent the current Track.
+        /// </summary>
+        public RectTransform[] MapPath;
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Map/MapTrack.cs]:
```suggestion
        public RectTransform[] mapPath;
```

**Review by lucasvdiepen — CHANGES_REQUESTED** [review]

**Review by PiterGroot — APPROVED** [review]

**Review by lucasvdiepen — APPROVED** [review]

**Pull request merged**